Fix: Incorrect duration metric recording incorrect values

novahe opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
in Go, arguments to deferred function calls are evaluated immediately
when the defer statement is executed, not when the deferred function
runs. This caused SinceInSeconds(startTime) to be computed at the
start of request handling instead of at the end.
```golang
package main

import (
	"fmt"
	"time"
)

func main() {
	start := time.Now()

	// --- 1. ❌ INCORRECT WAY: Immediate Argument Evaluation ---
	// The function time.Since(start) is an ARGUMENT to fmt.Println.
	// Go's defer rule states that arguments are evaluated IMMEDIATELY
	// when the defer statement is executed.
	// Result: Reports a time duration close to zero.
	defer fmt.Println("❌ Incorrect Value (Immediate Evaluation):", time.Since(start)) 

	// --- 2. ✅ CORRECT WAY: Using a Closure/Anonymous Function ---
	// The deferred function is now the anonymous function itself (func() {}).
	// The code inside the closure (time.Since(start)) is NOT evaluated
	// until the function exits and the deferred call is actually made.
	// Result: Reports the true duration of the main function's execution.
	defer func() {
		fmt.Println("✅ Correct Value (Delayed Evaluation):", time.Since(start))
	}()

	fmt.Println("Main function execution STARTING...")

	// Simulate a time-consuming operation (e.g., an HTTP request or database query)
	time.Sleep(100 * time.Millisecond) 

	fmt.Println("Main function execution FINISHED.")
	// When the function exits, deferred statements are executed (LIFO).
}
```
output:
```
Main function execution STARTING...
Main function execution FINISHED.
✅ Correct Value (Delayed Evaluation): 100.897289ms
❌ Incorrect Value (Immediate Evaluation): 231ns
```
#### Which issue(s) this PR is related to:
Fixes #135662
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #135662
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug where the event_handling_duration_seconds/preemption_goroutines_duration_seconds/run_podsandbox_duration_seconds/store_schedule_results_duration_seconds metric was recording
near-zero latency values instead of actual value.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

macsko commented on pkg/kubelet/kuberuntime/instrumented_services.go:
```diff
@@ -181,7 +181,9 @@ func (in instrumentedRuntimeService) RunPodSandbox(ctx context.Context, config *
 	startTime := time.Now()
 	defer recordOperation(operation, startTime)
-	defer metrics.RunPodSandboxDuration.WithLabelValues(runtimeHandler).Observe(metrics.SinceInSeconds(startTime))
+	defer func() {
```
Could you add a comment to each function to make sure noone will try to "optimize"?

macsko commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -53,7 +53,9 @@ import (
 	evt := fwk.ClusterEvent{Resource: fwk.Node, ActionType: fwk.Add}
 	start := time.Now()
-	defer metrics.EventHandlingLatency.WithLabelValues(evt.Label()).Observe(metrics.SinceInSeconds(start))
+	defer func() {
```
Can we somehow add detection of such problems to the linter?

macsko commented:
/ok-to-test

dgrisonnet commented:
@novahe good job finding these :+1:

Did you scan the whole codebase? Or could there be some other instances of this problem besides the one fixed in https://github.com/kubernetes/kubernetes/pull/135663?

/lgtm

dgrisonnet commented:
@novahe could you please pick the changes from https://github.com/kubernetes/kubernetes/pull/135663? I'll close my PR in favor of yours.

novahe commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -100,7 +102,9 @@ func (sched *Scheduler) updateNodeInCache(oldObj, newObj interface{}) {
 	start := time.Now()
-	defer metrics.EventHandlingLatency.WithLabelValues(evt.Label()).Observe(metrics.SinceInSeconds(start))
+	defer func() {
+		metrics.EventHandlingLatency.WithLabelValues(evt.Label()).Observe(metrics.SinceInSeconds(start))
```
Great point. I've refactored the code to use a new API that accepts startTime directly. This is much cleaner and avoids the potential pitfalls of defer evaluation.
```golang
// ObserveSince returns a function that observes the duration since the given start time.
// This is intended to be used with defer to measure the time elapsed since a given point:
//
//	start := time.Now()
//	defer metricVec.ObserveSince(start, "label1", "label2")()
//
// The returned function must be called (typically via defer) to record the observation.
// This pattern avoids the common pitfall where arguments to deferred functions are evaluated
// immediately, which would result in incorrect timing measurements.
func (v *HistogramVec) ObserveSince(start time.Time, lvs ...string) func() {
	return func() {
		v.WithLabelValues(lvs...).Observe(time.Since(start).Seconds())
	}
}
```

dgrisonnet committed:
Fix defer duration metric recording bug

Defer statements were incorrectly recording near-zero duration values (~2
microseconds) instead of actual elapsed time because arguments to deferred
functions are evaluated immediately when the defer statement is executed.

Solution: Added HistogramVec.ObserveSince() API that properly handles
deferred duration measurement, and replaced all problematic patterns.

The new API pattern:
  defer metric.ObserveSince(time.Now(), labels)()

Also deprecated SinceInSeconds() with warnings about incorrect defer usage.

Fixes: Incorrect HTTP request duration metrics in kubelet

dgrisonnet reviewed (approved):
/lgtm
/approve

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

novahe commented:
Resolve code conflicts, PTAL @SergeyKanzhelev @dgrisonnet

SergeyKanzhelev commented:
/lgtm
